src/grain/ledger.js

```diff
@@ -0,0 +1,220 @@
+  // involved at all (e.g. as a recipient or sender).
+  // Only exact address matches are included (no prefix matching).
+  // If null, all addresses are included.
+  +address: NodeAddressT | null,
+  // The order to return results in (ASCENDING or DESCENDING).
+  // Defaults to ASCENDING.
+  +eventOrder: EventOrderT,
+  // The sorting strategy to use (by TIMESTAMP or by AMOUNT).
+  // Defaults to TIMESTAMP.
+  +eventSort: EventSortT,
+|};
+
+export const DEFAULT_EVENTS_OPTIONS = deepFreeze({
+  address: null,
+  eventOrder: EventOrder.ASCENDING,
+  eventSort: EventSort.TIMESTAMP,
+});
+
+/**
+ * Models a transfer of grain from a sender to recipient.
+ * A grainholder may transfer grain to themself (which is a no-op).
+ */
+export type Transfer = {|
+  +type: "TRANSFER",
+  +version: string,
+  +sender: NodeAddressT,
+  +recipient: NodeAddressT,
+  +amount: Grain,
+  +timestampMs: number,
+|};
+
+/**
+ * Models a participant recieving newly minted grain.
+ */
+export type Distribution = {|
+  +type: "DISTRIBUTION",
+  +version: string,
+  +recipient: NodeAddressT,
+  +amount: Grain,
+  +timestampMs: number,
+|};
+
+export type LedgerEvent = Distribution | Transfer;
+
+/**
+ * A ledger tracks balances, earnings, and the event history of participants.
+ *
+ * We may refactor the `events` to be an async generator in the future,
+ * when we stop storing all the events in memory.
+ */
+export interface Ledger {
+  /**
+   * Stores the current grain balance of each address.
+   */
+  balances(): Map<NodeAddressT, Grain>;
+  /**
+   * Stores the lifetime earnings of each address.
+   *
+   * Necessary for computing which participants have been underpaid.
+   */
+  earnings(): Map<NodeAddressT, Grain>;
+  /**
+   * Retrieve the events.
+   *
+   * Can be filtered and sorted based on the
+   * EventsOptions, see that type for documentation.
+   */
+  events(?EventsOptions): $ReadOnlyArray<LedgerEvent>;
+}
+
+export class InMemoryLedger implements Ledger {
+  _balances: Map<NodeAddressT, Grain>;
+  _earnings: Map<NodeAddressT, Grain>;
+  _events: $ReadOnlyArray<LedgerEvent>;
+
+  /**
+   * Construct a Ledger from an array of in-memory events.
+   *
+   * The events must be in timestamp sorted order.
+   */
+  constructor(events: $ReadOnlyArray<LedgerEvent>) {
+    this._balances = new Map();
+    this._earnings = new Map();
+    this._events = events;
+    let lastTimestamp = -Infinity;
+    for (const e of events) {
+      if (e.timestampMs < lastTimestamp) {
+        throw new Error(
+          `event timestamps out of order: encountered ${e.timestampMs} after ${lastTimestamp}`
+        );
+      }
+      switch (e.type) {
+        case "DISTRIBUTION":
+          this._processDistribution((e: Distribution));
+          break;
+        case "TRANSFER":
+          this._processTransfer((e: Transfer));
+          break;
+        default:
+          throw new Error(`Unsupported event type: ${(e.type: empty)}`);
+      }
```

Comment:
Nit: generally think logic heavy constructors are a smell.
Here I would think it's a small improvement to have one of
```js
// Static (preferably pure) factory function
InMemoryLedger.from = ($ReadOnlyArray<LedgerEvent>) => InMemoryLedger

// Append / replay method to process events
const ledger = new InMemoryLedger();
ledger.replay(events);
```

Reply:
I've heard this a lot, I'm not really sure what the problem is with having logic in the constructor though. From an implementation perspective, I imagine re-writing the Ledger so the constructor takes an opaque bundle of fully-computed arguments, or doing your replay approach. But the replay approach invites questions about whether it's OK to replay multiple sequences of events on top of each other, which needs more testing, so overall the change feels a bit annoying.

If you feel strongly about it I can make one of these changes, but I'm kind of inclined to just merge as-is.

Reply:
Update: I decided to make the `InMemoryLedger` class a non-exported class, and just have a pure function which produces a fully constructed one. I believe this obviates concerns about having the logic-heavy constructor (it's now a pure implementation detail, we don't have to worry about others trying to inherit from it or something), without needing a rewrite.